### PR TITLE
Fix TS and ESLint errors in weeklyCheckoutFormGifting.tsx

### PR DESCRIPTION
**Pull request by GHaberis**
## What are you doing in this PR?

This PR addresses most of the TS/ESLint in `/weeklyCheckoutFormGifting.tsx`. 

This drew heavy influence from https://github.com/guardian/support-frontend/pull/3314 which addressed similar issues in `digitalCheckoutFormGift.tsx`.

High level overview:
- Rearrange and import order 
- Replaced explicit `PropTypes` definition as we can directly infer prop types from the `mapStateToProps`/`mapDispatchToProps` components, rather than having to duplicate type information. This meant I could also remove quite a few `Type` imports.

[Trello](https://trello.com/c/Gkx2iWvT/146-assets-pages-weekly-subscription-checkout-components-weeklycheckoutformgiftingtsx)

### CODE REVIEW

File: support-frontend/assets/pages/weekly-subscription-checkout/components/weeklyCheckoutFormGifting.tsx [hunk 1]
```diff
@@ -130,34 +95,37 @@ function mapStateToProps(state: WithDeliveryCheckoutState) {
 		isTestUser: state.page.checkout.isTestUser,
 		country: state.common.internationalisation.countryId,
 		csrf: state.page.csrf,
-		currencyId: currencyFromCountryCode(deliveryAddress.fields.country),
+		currencyId:
```
Comment (GHaberis) [inline support-frontend/assets/pages/weekly-subscription-checkout/components/weeklyCheckoutFormGifting.tsx:98]: `currencyId` cannot be `null`, however `currencyFromCountryCode` can return null if the country code currency lookup can't find a matching currency, in this case I've opted to fallback to the `defaultCurrency` for the region.

File: support-frontend/assets/pages/weekly-subscription-checkout/components/weeklyCheckoutFormGifting.tsx [hunk 2]
```diff
@@ -130,34 +95,37 @@ function mapStateToProps(state: WithDeliveryCheckoutState) {
...
 			(
 				dispatch: Dispatch<Action>,
 				getState: () => WithDeliveryCheckoutState,
 			) => {
 				fetchAndStoreUserType(email)(dispatch, getState);
 			},
 		formIsValid:
-			() =>
-			(dispatch: Dispatch<Action>, getState: () => WithDeliveryCheckoutState) =>
+			() => (_: Dispatch<Action>, getState: () => WithDeliveryCheckoutState) =>
```
Comment (GHaberis) [inline support-frontend/assets/pages/weekly-subscription-checkout/components/weeklyCheckoutFormGifting.tsx:119]: the first argument (FKA as `dispatch`) is not used, an underscore indicates that in Typescript and means we don't get an unused argument error.

File: support-frontend/assets/pages/weekly-subscription-checkout/components/weeklyCheckoutFormGifting.tsx [hunk 3]
```diff
@@ -256,44 +232,35 @@ function WeeklyCheckoutFormGifting(props: PropTypes) {
...
+								label="Gift delivery date"
 							>
 								{days.map((day) => {
 									const [userDate, machineDate] = [
 										formatUserDate(day),
 										formatMachineDate(day),
 									];
-									const hideDate = machineDate.endsWith('-12-25');
-
-									// Don't render input if Christmas day
```
Comment (tomrf1) [inline support-frontend/assets/pages/weekly-subscription-checkout/components/weeklyCheckoutFormGifting.tsx:293]: no longer needed?

File: support-frontend/assets/pages/weekly-subscription-checkout/components/weeklyCheckoutFormGifting.tsx [hunk 4]
```diff
@@ -180,17 +148,24 @@ function mapDispatchToProps() {
 	};
 }
 
+const connector = connect(mapStateToProps, mapDispatchToProps());
+
+// ----- Types ----- //
+type PropTypes = ConnectedProps<typeof connector>;
```
Comment (tomrf1) [inline support-frontend/assets/pages/weekly-subscription-checkout/components/weeklyCheckoutFormGifting.tsx:154]: nice!

File: support-frontend/assets/pages/weekly-subscription-checkout/components/weeklyCheckoutFormGifting.tsx (hunk 3 above)
Comment (GHaberis) [reply support-frontend/assets/pages/weekly-subscription-checkout/components/weeklyCheckoutFormGifting.tsx:293]: It was a short-term fix to address 25th December being on a Friday and therefore listed as an eligible delivery date in 2020. The next time 25th December is on a Friday is 2026, hopefully by then we'll have implemented a better solution! There is a ticket to use the fulfillment date calculator developed by Retention, card to do that is here: https://trello.com/c/bovZtRnp/151-utilise-fulfilment-date-calculator-to-centralise-logic-for-fulfilment-dates.

File: support-frontend/ts-error-history.csv [hunk 5]
```diff
@@ -35,4 +35,8 @@
 1637070377000,1135,596e3e86b5e80d3349c6e611a4cb9d7d895a4592
 1637080641000,1116,6a596404e0e8259143074451cbd1381aec01f641
 1637139236000,1116,e92ec73f53bb872c9cd5080884b0a7faa6f63681
+<<<<<<< HEAD
 1637151796000,1116,HEAD
+=======
+1637161943000,1142,HEAD
+>>>>>>> 5e2c7d0f6 (fix TS and ESlint errors)
```
Comment (TomPretty) [inline support-frontend/ts-error-history.csv:42]: I think we'll want to fix this one!

File: support-frontend/typescript-errors.json [hunk 6]
```diff
@@ -4085,6 +4044,14 @@
...
             "path": "assets/helpers/tracking/googleTagManager.ts",
             "message": "Could not find a declaration file for module 'uuid'. 'support-frontend/node_modules/uuid/index.js' implicitly has an 'any' type."
+        },
+        {
+            "path": "assets/pages/contributions-landing/components/ContributionFormContainer.tsx",
+            "message": "Could not find a declaration file for module 'react-router-dom'. 'support-frontend/node_modules/react-router-dom/index.js' implicitly has an 'any' type."
+        },
+        {
+            "path": "assets/pages/contributions-landing/contributionsLanding.tsx",
+            "message": "Could not find a declaration file for module 'react-router-dom'. 'support-frontend/node_modules/react-router-dom/index.js' implicitly has an 'any' type."
```
Comment (TomPretty) [inline support-frontend/typescript-errors.json:4054]: Should be able to fix this by running `yarn` to install new declarations

File: support-frontend/typescript-errors.json [hunk 7]
```diff
@@ -788,6 +787,14 @@
...
             "message": "Parameter 'result' implicitly has an 'any' type."
         },
+        {
+            "path": "assets/pages/contributions-landing/contributionsLanding.tsx",
+            "message": "Parameter 'props' implicitly has an 'any' type."
+        },
+        {
+            "path": "assets/pages/contributions-landing/contributionsLanding.tsx",
+            "message": "Parameter 'props' implicitly has an 'any' type."
+        },
```
Comment (TomPretty) [inline support-frontend/typescript-errors.json:797]: I'm not quite sure why these have been introduced :/ maybe running yarn will fix them too?

File: support-frontend/assets/pages/weekly-subscription-checkout/components/weeklyCheckoutFormGifting.tsx (hunk 4 above)
Comment (TomPretty) [reply support-frontend/assets/pages/weekly-subscription-checkout/components/weeklyCheckoutFormGifting.tsx:154]: I guess I'm slightly mixed on this one - it's nice to reduce code, but having an explicit definition of prop types is also quite nice for quickly checking what type everything is

File: support-frontend/ts-error-history.csv (hunk 5 above)
Comment (GHaberis) [reply support-frontend/ts-error-history.csv]: Huh, though I'd rebased that. Fixed it now.

File: support-frontend/assets/pages/weekly-subscription-checkout/components/weeklyCheckoutFormGifting.tsx (hunk 4 above)
Comment (GHaberis) [reply support-frontend/assets/pages/weekly-subscription-checkout/components/weeklyCheckoutFormGifting.tsx:154]: @TomPretty I see where you're coming from, although with my IDE (Visual Studio) it's possible to view the properties and types of `PropTypes` still.

**Pull request merged**